### DISCUSSION TITLE
How to use Alembic with Postgres?

### DISCUSSION THREAD

**Discussion by InAnYan** [sqlalchemy/alembic#1501] [Usage Questions]
This drives me crazy, I tried several tutorials, but it doesn't work.

None of the commands work, I get this error:
```
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.InvalidSchemaName) ОШИБКА:  схема для создания объектов не выбрана
LINE 2: CREATE TABLE alembic_version (
                     ^

[SQL:
CREATE TABLE alembic_version (
        version_num VARCHAR(32) NOT NULL,
        CONSTRAINT alembic_version_pkc PRIMARY KEY (version_num)
)

]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

`схема для создания объектов не выбрана` means `schema for creating objects is not chosen`.

What should I do?

**Comment by InAnYan** [comment]
Okay, another tutorial: https://medium.com/@johnidouglasmarangon/using-migrations-in-python-sqlalchemy-with-alembic-docker-solution-bd79b219d6a seems to work.

But, why I got the migration:
```python
"""empty message

Revision ID: 0b1d71871aa2
Revises: 
Create Date: 2024-07-07 12:06:32.045538

"""
from typing import Sequence, Union

from alembic import op
import sqlalchemy as sa


# revision identifiers, used by Alembic.
revision: str = '0b1d71871aa2'
down_revision: Union[str, None] = None
branch_labels: Union[str, Sequence[str], None] = None
depends_on: Union[str, Sequence[str], None] = None


def upgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.drop_table('alembic_version')
    # ### end Alembic commands ###


def downgrade() -> None:
    # ### commands auto generated by Alembic - please adjust! ###
    op.create_table('alembic_version',
    sa.Column('version_num', sa.VARCHAR(length=32), autoincrement=False, nullable=False),
    sa.PrimaryKeyConstraint('version_num', name='alembic_version_pkc')
    )
    # ### end Alembic commands ###
```

1. Why upgrade and downgrade doing the opposite thing?
2. ~Why my tables are not added?~

UPDATE: question 2 solved

**Reply by InAnYan** [reply]
Ah, sorry, I see, for question 2 I have to import all models

**Comment by InAnYan** [comment]
after some mangling with `alembic_version`:
1. I've swapped dropping and creating
2. Then I completely deleted it

It works now, but still, what is a proper way to do it? I'm confused a little bit

**Comment by CaselIT** [comment]
Hi,

You seem to have a strange setup regarding the postgresql `search_path` that prevents you from creating tables without a schema.
Either fix that issue or specify a schema in the env so that the alembic version table is generated in the correct schema using https://alembic.sqlalchemy.org/en/latest/api/runtime.html#alembic.runtime.environment.EnvironmentContext.configure.params.version_table_schema
